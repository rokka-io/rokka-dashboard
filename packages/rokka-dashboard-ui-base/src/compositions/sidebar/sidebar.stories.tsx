import { action } from '@storybook/addon-actions';
import { boolean, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import { SearchInput } from '../../components';
import { AddIcon, LiPosRelative, ShowOnActive, SidebarLink, SidebarLinkIcon } from '../../elements';
import { SidebarStacksList } from '../../elements/sidebarStacksList/sidebarStacksList';
import { Sidebar } from './sidebar';

storiesOf('Compositions', module)
  .addDecorator(story => <div style={{ minHeight: '100vh', transform: 'translate(0)' }}>{story()}</div>)
  .add('Sidebar', () => (
    <Sidebar active={boolean('active', false)}>
      <li>
        <SidebarLink>Test</SidebarLink>
      </li>
      <LiPosRelative>
        <SidebarLink>Stacks</SidebarLink>
        <SidebarLinkIcon>
          <AddIcon height="24px" width="24px" />
        </SidebarLinkIcon>
        <ShowOnActive active={boolean('stacks list active', true)}>
          <SearchInput onChange={action('onChange')} placeholder={text('placeholder', 'Search...')} />
          <SidebarStacksList>
            <li>
              <SidebarLink small={true}>A stack</SidebarLink>
            </li>
            <li>
              <SidebarLink small={true}>2nd stack</SidebarLink>
            </li>
          </SidebarStacksList>
        </ShowOnActive>
      </LiPosRelative>
    </Sidebar>
  ));
