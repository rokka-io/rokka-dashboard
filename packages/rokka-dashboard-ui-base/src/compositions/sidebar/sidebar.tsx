import React, { FunctionComponent, ReactNode } from 'react';
import styled from 'styled-components';
import { PoweredBy } from '../../components/poweredBy/poweredBy';
import { colors } from '../../identity/colors/colors';
import { media } from '../../identity/media/media';
import { spaces } from '../../identity/spaces/spaces';
import { zIndex } from '../../identity/zIndex';

interface SidebarProps {
  active?: boolean;
  children: ReactNode;
}

export const Sidebar: FunctionComponent<SidebarProps> = ({ active = false, children }) => (
  <StyledNav active={active}>
    <StyledUl>{children}</StyledUl>
    <PoweredBy />
  </StyledNav>
);

interface StyledNavProps {
  active?: boolean;
}

const StyledNav = styled.nav<StyledNavProps>`
  position: fixed;
  left: 0;
  bottom: 0;
  top: ${spaces.xlarge};
  width: 13.75rem;
  background: ${colors.gray.sidebar};
  color: ${colors.gray.light};
  transform: translateX(-13.75rem);
  transition: transform .4s ease;
  z-index: ${zIndex.sidebar};

  ${media.desktop`
    transform: translateX(0);
  `}

  ${media.mobile`
    transform: translateX(-100%);
  `}

  ${({ active }) =>
    active &&
    `
    transform: translateX(0);
    transition: transform .4s ease;

    ${media.mobile`
      width: 100%;
    `}
  `}
`;

const StyledUl = styled.ul`
  margin: 0;
  padding: 0;

  > li {
    list-style: none;
  }
`;
