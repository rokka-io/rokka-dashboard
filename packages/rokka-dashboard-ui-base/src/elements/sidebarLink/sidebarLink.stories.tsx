import { boolean, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import { MemoryRouter as Router } from 'react-router';
import { colors } from '../../identity/colors/colors';
import { AddIcon } from '../icon/icon';
import { SidebarLink, SidebarLinkIcon } from './sidebarLink';

storiesOf('Elements / Sidebar', module)
  .addDecorator(story => (
    <Router>
      <div style={{ position: 'relative', background: colors.gray.sidebar, height: '55px' }}>{story()}</div>
    </Router>
  ))
  .add('SidebarLink', () => (
    <SidebarLink to="#" active={boolean('active', false)} small={boolean('small', false)}>
      {text('text', 'Home')}
    </SidebarLink>
  ))
  .add('SidebarLinkIcon', () => (
    <>
      <SidebarLink to="#" active={boolean('active', false)} small={boolean('small', false)}>
        {text('text', 'New stack')}
      </SidebarLink>
      <SidebarLinkIcon to="#" active={boolean('active', false)} small={boolean('small', false)}>
        <AddIcon height="24px" width="24px" />
      </SidebarLinkIcon>
    </>
  ));
