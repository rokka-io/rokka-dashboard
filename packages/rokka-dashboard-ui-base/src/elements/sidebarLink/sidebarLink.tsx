import React, { FunctionComponent, ReactNode } from 'react';
import styled, { css } from 'styled-components';
import { colors } from '../../identity/colors/colors';
import { spaces } from '../../identity/spaces/spaces';
import { fonts } from '../../identity/typography';

interface SidebarLinkProps {
  /** Where the link leads to */
  href?: string;
  /** Is link currently active */
  active?: boolean;
  /** Is sub navigation item */
  sub?: boolean;
  /** Content of link */
  children: ReactNode;
}

export const SidebarLink: FunctionComponent<SidebarLinkProps> = ({ children, active, sub, href }) => (
  <StyledLink href={href} active={active} sub={sub}>
    {children}
  </StyledLink>
);

export const SidebarLinkIcon: FunctionComponent<SidebarLinkProps> = ({ children, active, sub, href }) => (
  <StyledLinkIcon href={href} active={active} sub={sub}>
    {children}
  </StyledLinkIcon>
);

interface StyledLinkProps {
  /** Is link currently active */
  active?: boolean;
  /** Is sub navigation item */
  sub?: boolean;
}

const ActiveStyle = css`
  transition: background 0.2s ease, border 0.2s ease;
  background: rgba(255, 255, 255, 0.05);
  border-color: ${colors.brand.light};
`;

const StyledLink = styled.a<StyledLinkProps>`
  font-size: ${({ sub }) => (sub ? fonts.Sizes.small : fonts.Sizes.large)};
  padding: ${({ sub }) => (sub ? `${spaces.small} ${spaces.medium}` : spaces.medium)};
  border-left: 8px solid transparent;
  display: block;
  color: ${colors.tints.white};
  text-decoration: none;
  transition: background 0.2s ease, border 0.2s ease;

  ${({ active }) => active && ActiveStyle}

  :hover {
    ${ActiveStyle}
  }
`;

const ActiveStyleIcon = css`
  transition: 0.3s ease color;
  color: ${colors.brand.primary};
`;

const StyledLinkIcon = styled.a<StyledLinkProps>`
  transition: 0.3s ease color;
  position: absolute;
  right: ${({ sub }) => (sub ? spaces.small : spaces.medium)};
  top: ${({ sub }) => (sub ? spaces.small : spaces.medium)};
  color: ${colors.tints.white};

  ${({ active }) => active && ActiveStyleIcon}

  :hover {
    ${ActiveStyleIcon}
  }
`;
