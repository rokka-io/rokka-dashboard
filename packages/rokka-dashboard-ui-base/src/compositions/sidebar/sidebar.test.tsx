import { shallow } from 'enzyme';
import React from 'react';

import { SidebarLink } from '../../elements/sidebarLink/sidebarLink';
import { Sidebar } from './sidebar';

describe('Compositions', () => {
  describe('Sidebar', () => {
    it('renders a sidebar correctly.', () => {
      const component = shallow(
        <Sidebar>
          <li>
            <SidebarLink>Test</SidebarLink>
          </li>
        </Sidebar>
      );

      expect(component).toMatchSnapshot();
    });
    it('renders an active sidebar correctly', () => {
      const component = shallow(
        <Sidebar active={true}>
          <li>
            <SidebarLink>Test</SidebarLink>
          </li>
        </Sidebar>
      );

      expect(component).toMatchSnapshot();
    });
  });
});
