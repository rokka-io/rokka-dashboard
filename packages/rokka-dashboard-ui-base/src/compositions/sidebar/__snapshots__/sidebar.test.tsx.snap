// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compositions Sidebar renders a sidebar correctly. 1`] = `
<styled.nav
  active={false}
>
  <styled.ul>
    <li>
      <Component
        to="#"
      >
        Test
      </Component>
    </li>
  </styled.ul>
  <Component />
</styled.nav>
`;

exports[`Compositions Sidebar renders an active sidebar correctly 1`] = `
<styled.nav
  active={true}
>
  <styled.ul>
    <li>
      <Component
        to="#"
      >
        Test
      </Component>
    </li>
  </styled.ul>
  <Component />
</styled.nav>
`;
